java/ql/src/semmle/code/java/dataflow/internal/rangeanalysis/SignAnalysisSpecific.qll

```diff
@@ -0,0 +1,235 @@
+    exists(ReflectiveFieldAccess rfa | rfa.inferAccessedField() = f)
+    or
+    if f.fromSource()
+    then not exists(f.getInitializer()) and result = TZero()
+    else
+      if f instanceof ArrayLengthField
+      then result != TNeg()
+      else
+        if f.hasName("MAX_VALUE")
+        then result = TPos()
+        else
+          if f.hasName("MIN_VALUE")
+          then result = TNeg()
+          else any()
+  }
+
+  Sign specificSubExprSign(Expr e) {
+    result = exprSign(e.(AssignExpr).getSource())
+    or
+    result = exprSign(e.(PlusExpr).getExpr())
+    or
+    result = exprSign(e.(PostIncExpr).getExpr())
+    or
+    result = exprSign(e.(PostDecExpr).getExpr())
+    or
+    result = exprSign(e.(PreIncExpr).getExpr()).inc()
+    or
+    result = exprSign(e.(PreDecExpr).getExpr()).dec()
+    or
+    result = exprSign(e.(MinusExpr).getExpr()).neg()
+    or
+    result = exprSign(e.(BitNotExpr).getExpr()).bitnot()
+    or
+    exists(DivExpr div |
+      div = e and
+      result = exprSign(div.getLeftOperand()) and
+      result != TZero()
+    |
+      div.getRightOperand().(FloatingPointLiteral).getValue().toFloat() = 0 or
+      div.getRightOperand().(DoubleLiteral).getValue().toFloat() = 0
+    )
+    or
+    exists(Sign s1, Sign s2 | binaryOpSigns(e, s1, s2) |
+      (e instanceof AssignAddExpr or e instanceof AddExpr) and
+      result = s1.add(s2)
+      or
+      (e instanceof AssignSubExpr or e instanceof SubExpr) and
+      result = s1.add(s2.neg())
+      or
+      (e instanceof AssignMulExpr or e instanceof MulExpr) and
+      result = s1.mul(s2)
+      or
+      (e instanceof AssignDivExpr or e instanceof DivExpr) and
+      result = s1.div(s2)
+      or
+      (e instanceof AssignRemExpr or e instanceof RemExpr) and
+      result = s1.rem(s2)
+      or
+      (e instanceof AssignAndExpr or e instanceof AndBitwiseExpr) and
+      result = s1.bitand(s2)
+      or
+      (e instanceof AssignOrExpr or e instanceof OrBitwiseExpr) and
+      result = s1.bitor(s2)
+      or
+      (e instanceof AssignXorExpr or e instanceof XorBitwiseExpr) and
+      result = s1.bitxor(s2)
+      or
+      (e instanceof AssignLShiftExpr or e instanceof LShiftExpr) and
+      result = s1.lshift(s2)
+      or
+      (e instanceof AssignRShiftExpr or e instanceof RShiftExpr) and
+      result = s1.rshift(s2)
+      or
+      (e instanceof AssignURShiftExpr or e instanceof URShiftExpr) and
+      result = s1.urshift(s2)
+    )
+    or
+    result = exprSign(e.(ChooseExpr).getAResultExpr())
+    or
+    result = exprSign(e.(CastExpr).getExpr())
+  }
+
+  private Sign binaryOpLhsSign(Expr e) {
+    result = exprSign(e.(BinaryExpr).getLeftOperand()) or
+    result = exprSign(e.(AssignOp).getDest())
+  }
+
+  private Sign binaryOpRhsSign(Expr e) {
+    result = exprSign(e.(BinaryExpr).getRightOperand()) or
+    result = exprSign(e.(AssignOp).getRhs())
+  }
+
+  pragma[noinline]
+  private predicate binaryOpSigns(Expr e, Sign lhs, Sign rhs) {
```

Comment:
This could be moved to the common bit, but the gain would be minimal. I think it's better to keep it together with `binaryOpLhsSign` and `binaryOpRhsSign`.